Apache Thrift language binding for Dart

markerickson-wf opened the pull request:
IN PROGRESS

Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Currently basic functionality is implemented and the tutorial is working.  Integration tests will be written before a PR to Apache is made.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Added t_dart_generator.cc and add as a target for Make.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Cleanup some issues in the generator found while analyzing generated code.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Initial port of library code, except for transport implementations.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Minor tweaks to the dart generator

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Added binary and JSON protocols.  Added WebSocket transport.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Switch service generation of service server code to opt-in instead of opt-out.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Added THttpTransport and a did little refactoring.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Added BufferedTransport and FramedTransport, with some refactoring.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Added MultiplexedProtocol.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Ran dartfmt.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Optimizations when using List.length since the base implementation iterates over elements in the getter.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Fix BufferedTransport.flush to be able to read from the write buffer.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Fix typo in Dart SDK requirement

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Add pubspec.yaml to generated files.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Fix generation problems exposed by tutorial specification, related to inheritance and constants.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Initialize seqid and increment when client writes message begin.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

In progress.  Initial stab at tutorial implementation, with known issues.  Changed socket interface for open / close to be async.  Added support for sockets that are listening for messages (server).

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Fix bugs in binary protocol and json protocol.  Use base 64 encoding for socket and http transports.  Implement the remainder of the tutorial.

markerickson-wf committed:
Merge branch 'master' into thrift-dart

evanweible-wf commented on lib/dart/lib/src/browser/t_web_socket.dart:
```diff
@@ -0,0 +1,138 @@
+
+  WebSocket _socket;
+
+  bool get isOpen => _socket != null && _socket.readyState == WebSocket.OPEN;
```
Since you're requiring dart 1.12.0, you can use null-aware operators here:

``` dart
bool get isOpen => _socket?.readyState == WebSocket.OPEN;
```

markerickson-wf replied:
I was holding off on those until tooling catches up and supports them.  But they will be nice!

evanweible-wf commented on lib/dart/lib/src/browser/t_web_socket.dart:
```diff
@@ -0,0 +1,138 @@
+
+  Future close() async {
+    if (_socket != null) {
+      _socket.close();
```
This method is async but you're not awaiting anything

markerickson-wf replied:
I'm using `async` to return the Future implicitly, to fulfill the interface.

stevenosborne-wf replied:
Maybe return `_socket.onClose.first`.

markerickson-wf replied:
Good idea.  I addressed this in https://github.com/markerickson-wf/thrift/commit/cd83635f4ae688d2ae65d474790a213294576d15.

evanweible-wf commented on lib/dart/lib/src/browser/t_web_socket.dart:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    if (!_completers.isEmpty) {
+      _completers.removeAt(0).complete(data);
```
Just to make sure I understand this correctly, is this assuming that every request sent to the server will be accompanied by a response? Is it actually guaranteed that they will come back in the same order? Or do they need to be mapped to an ID?

markerickson-wf replied:
Yeah good catch it is assuming order, and for a client it assumes there will be a response.  This is actually what the JavaScript implementation does, but I had the same concerns when porting it over.

https://github.com/apache/thrift/blob/master/lib/js/src/thrift.js#L577

I think using the seqid would be preferable, but it seems that would require knowledge of the protocol being used to decode that, which the transport doesn't have.  We should think about this some more.

evanweible-wf commented on lib/dart/lib/src/browser/t_web_socket.dart:
```diff
@@ -0,0 +1,138 @@
+  void _onClose(CloseEvent event) {
+    _socket = null;
+    _completers.clear();
+    _requests.clear();
```
Is it desirable to clear out pending requests when the socket closes? Would we ever need to know which requests didn't complete?

markerickson-wf replied:
Good catch.  We should probably complete any remaining completers (both those in the active queue and those in requests) with an error before clearing.

evanweible-wf commented on lib/dart/lib/src/protocol/t_json_protocol.dart:
```diff
@@ -0,0 +1,733 @@
+  }
+
+  int _hexChar(int byte) {
+    return byte.toRadixString(16).codeUnitAt(0);
```
`=>`

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Make pubspec dependencies more consistent and restrictive.

evanweible-wf commented on lib/dart/lib/src/transport/t_socket_transport.dart:
```diff
@@ -0,0 +1,77 @@
+  TSocketTransport(this.socket, {isListening: false})
+      : this.isListening = isListening {
+    if (socket == null) {
+      throw new ArgumentError.notNull("socket");
```
single quotes

evanweible-wf commented on tutorial/dart/client/pubspec.yaml:
```diff
@@ -0,0 +1,17 @@
+homepage: https://github.com/apache/thrift
+
+environment:
+  sdk: '>=1.0.0'
```
I read above that this requires dart >= 1.12.0, should that be reflected here?

evanweible-wf replied:
also, double quotes

evanweible-wf commented on tutorial/dart/client/pubspec.yaml:
```diff
@@ -0,0 +1,17 @@
+  sdk: '>=1.0.0'
+
+dependencies:
+  browser: '>=0.10.0'
```
double quotes

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Switch uses of List<int> to Uint8List to be more explicit.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Complete with error when a socket is closed.  Minor code cleanup.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Writing unit tests.  A little code cleanup.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

JSON protocol unit tests.  Changed binary type to Uint8List.  Cleaned up tutorial shell script.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Protocol and transport unit tests.

stevenosborne-wf commented on lib/dart/lib/src/browser/t_web_socket.dart:
```diff
@@ -0,0 +1,148 @@
+
+    _socket = new WebSocket(url.toString());
+    _socket.onError.listen(_onError);
+    _socket.onOpen.listen(_onOpen);
```
Presumably, we wouldn't want `open` to complete until `onOpen` completes. Could you return `_socket.onOpen.first`?

markerickson-wf replied:
Good idea.  I addressed this in https://github.com/markerickson-wf/thrift/commit/cd83635f4ae688d2ae65d474790a213294576d15.

stevenosborne-wf commented:
First pass done. Looks great Mark!

tylertreat-wf commented on lib/dart/lib/src/transport/t_buffered_transport.dart:
```diff
@@ -0,0 +1,98 @@
+
+/// Buffered implementation of [TTransport].
+class TBufferedTransport extends TTransport {
+  final List<int> _writeBuffer = [];
```
Is there a reason you use `List<int>` here instead of `Uint8List`?

markerickson-wf replied:
Uint8List is fixed length.  I need an internal buffer that is "growable".

tylertreat-wf commented:
Got through most of the c++ code and it lgtm (though it's been a while since I've done any c++). Also went through the dart code and it looks good based on my primitive knowledge of dart.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Move completer logic from WebSocket implementations to transport.  Split TSocketTransport into TClientSocketTransport and TServerSocketTransport to cleanup some conditional logic.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Move Base 64 encoding out of transport and back into web_socket implementations.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Return Futures from open and close that actually reflect the state of the WebSocket.

Commit:
Merge pull request #2 from markerickson-wf/socket-transport

Cleanup TSocketTransport and related code

stevenosborne-wf commented:
+1

evanweible-wf commented on lib/dart/lib/src/browser/t_web_socket.dart:
```diff
@@ -0,0 +1,130 @@
+    _socket.onClose.listen(_onClose);
+    _socket.onMessage.listen(_onMessage);
+
+    return _socket.onOpen.first;
```
Are these socket streams broadcast or single subscription? If they're not broadcast, this will throw since you're listening to onOpen above and here.

markerickson-wf replied:
`onOpen` and `onClose` for an html `WebSocket` return a `Stream` from `EventStreamProvider.forTarget()`, which is documented to return a broadcast stream. 

https://api.dartlang.org/1.12.1/dart-html/EventStreamProvider/forTarget.html

johnryan-wf commented on lib/dart/lib/src/t_base.dart:
```diff
@@ -0,0 +1,37 @@
+
+part of thrift;
+
+abstract class TBase {
```
is this being used?

markerickson-wf replied:
It's used by generated code.

johnryan-wf commented on lib/dart/lib/src/transport/t_transport.dart:
```diff
@@ -0,0 +1,70 @@
+  /// Throws [TTransportError] if there was an error writing data
+  void writeAll(Uint8List buffer) {
+    write(buffer, 0, buffer.length);
+  }
```
(nitpick) I would expect this method to be defined after `write()` (read, readAll, write, writeAll)

markerickson-wf replied:
I agree, thanks.

johnryan-wf commented:
Looks great!  This is much appreciated!

evanweible-wf commented on lib/dart/lib/src/browser/t_web_socket.dart:
```diff
@@ -0,0 +1,130 @@
+
+  final List<Uint8List> _requests = [];
+
+  TWebSocket(this.url)
```
Seems a bit odd to me that the browser version expects a URL and the console version expects an instance of `dart:io.WebSocket`.

This also creates inconsistent behavior around `open()/close()` - as it is written now, the browser version could be reopened once closed since it instantiates a new underlying websocket, but the console version just assumes that the underlying websocket is open and adds that state to the state stream.

What is the goal for this socket class - should it be reopenable? Should it abstract the underlying websocket class away completely?

evanweible-wf replied:
I see that the use case for creating a TWebSocket from an instance of WebSocket is in doing so from an upgrading request where you already have the websocket. It still seems that the default constructor for `TWebSocket` is different dependent on whether you import the browser or console version. This seems more like the console version should have an additional constructor since it can support being a websocket server and connecting to a websocket as a client.

markerickson-wf replied:
That's a good observation.  I mainly implemented the console version to support running a server for the tutorial.  

I didn't finish a full TServer implementation (there is more to be done there), and I was a little uncertain as to how much time to invest on the server / console code, since that support is being left TODO.  I also don't have a good use case for a console app as a client, but I agree it could and probably should be supported.

Basically the question is whether to leave this as TODO or clean this up now.

evanweible-wf replied:
Ah I see, didn't realize there was a TServer interface. I'm fine with either

markerickson-wf replied:
We'll leave this as TODO.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Avoid using String.codeUnits as a Uint8List.  Cleanup bad type assignments in tutorial client.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Add TMessageReader and use in TClientSocketTransport to correlate requests and responses by seqid

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Fix typo in error message.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Add support for a timeout when waiting for a response.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Make TMessageReader not extend TTransport, move to internal class _TMessageReaderTransport

Commit:
Merge pull request #3 from markerickson-wf/socket-read-seqid

Correlate requests and responses by seqid

Commit:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Use dart_dev for code formatting, tests, coverage, and adding license.

markerickson-wf committed:
Create an Apache Thrift language binding for Dart (dartlang.org).
https://issues.apache.org/jira/browse/THRIFT-3299

Remove unused generator argument from build.sh

Commit:
Merge pull request #4 from evanweible-wf/dart-cleanup

Dart tooling and cleanup

markerickson-wf commented:
PR has been made to Apache - https://github.com/apache/thrift/pull/608

johnryan-wf commented:
:+1:
